overlord/configstate/configcore/certs.go

```diff
@@ -32,6 +33,221 @@ import (
+		if cert.IsDir() {
+			continue
+		}
+
+		// So what are we trying to achieve here? When provided with certificate directories
+		// they may or may not be symbolic links to the actual certificate file. To handle this
+		// correctly we evaluate the symbolic links and get the real path of the certificate, because
+		// we are not going to 'copy' or link to the symbolic link, instead we will recreate the symbolic
+		// link to the real path of the certificate.
+		certRealPath := filepath.Join(basePath, cert.Name())
+		if cert.Mode()&os.ModeSymlink != 0 {
+			resolvedPath, err := filepath.EvalSymlinks(certRealPath)
+			if err == nil {
+				certRealPath = resolvedPath
+			}
+		}
+
+		// If the file is not a symbolic link then Path and RealPath will be identical.
+		certObject := certificate{
+			Name:     cert.Name(),
+			Path:     filepath.Join(basePath, cert.Name()),
+			RealPath: certRealPath,
+		}
+		certsObjects = append(certsObjects, certObject)
+	}
+	return certsObjects, nil
+}
+
+// Filters out the certificates that are not allowed to be merged.
+func filterCerts(certs []certificate, blockedCerts []string) ([]certificate, error) {
+	filteredCerts := []certificate{}
+	for _, cert := range certs {
+		if isBlocked(cert, blockedCerts) {
+			continue
+		}
+		filteredCerts = append(filteredCerts, cert)
+	}
+	return filteredCerts, nil
+}
+
+func filterCertsInDirectory(dirPath string, blockedCerts []string) error {
+	existingCerts, err := getCertObjects(dirPath)
+	if err != nil {
+		return err
+	}
+
+	// Remove the certs that are marked blocked.
+	for _, cert := range existingCerts {
+		if isBlocked(cert, blockedCerts) {
+			err := os.Remove(cert.Path)
+			if err != nil {
+				return fmt.Errorf("cannot remove blocked certificate: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// Populates symbolic links in the output directory, to each certificate
+// provided.
+func installCerts(outputPath string, certs []certificate) error {
+	for _, cert := range certs {
+		// When updating existing certifications we want to make sure we don't
+		// overwrite the certification with itself.
+		certDestinationPath := filepath.Join(outputPath, cert.Name)
+		if certDestinationPath == cert.RealPath {
+			continue
+		}
+
+		// If the path exists, then remove it first to update the certificate
+		if _, err := os.Lstat(certDestinationPath); err == nil {
+			err := os.Remove(certDestinationPath)
+			if err != nil {
+				return fmt.Errorf("cannot remove existing certificate: %v", err)
+			}
+		}
+
+		// Create a link to the real path of the cert in the output path
+		err := os.Symlink(cert.RealPath, certDestinationPath)
+		if err != nil {
+			return fmt.Errorf("cannot copy store certificate: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// Generate the ca-certificates.crt to the output path
+// The ca-certificates.crt is a concatenation of all the certs in the
+// output path.
+func generateCACertificate(outputPath string) error {
+	certificates, err := getCertObjects(outputPath)
```

Comment:
If `ca-certificates.crt` is already in `outputPath`, because this is run a second time, would not `getCertObjects` also return `ca-certificates.crt` in the list? I would have expecting `generateCACertificate` to take as parameter a `[]certificate`.

Also renaming this function to something like `generateCABunle` would help comprehension. Or at least put an "s". Because we are not generating a certificate here, just a bundle.

Reply:
This is a really good question! Actually this will never happen, because before we start generating the ca-certificate we actually delete any blocked certificates (including ca-certificates.crt) in the output folder, this is to take care of the output folder being a folder that contains ssl-certificates already that we want to keep.

Take a look at line 233 (where filterCertsInDirectory is called)

Reply:
I see, that you do not non-symlinked files in `getCertObjects`. That is probably why that works.

Reply:
getCertsObjects actually returns files that are non-symlinks as well. But before calling getCertsObjects we have already deleted the ca-certificates.crt file from the directory, so it never appears in the list